agent.py

```diff
@@ -156,13 +157,133 @@ def __get_action(self, new_state):
+        with torch.no_grad():
+            action_q_vals = self.policy_net(observation.view(1, self.channels, self.grid_size, self.grid_size)) 
+        return torch.argmax(action_q_vals).item()
+
+    def initialize(self):
+        self.__experience_initial()
+
+    def update(self):
+        observation, reward, terminated, truncated, info = self.__experience_replay()
+        
+        states, new_states, actions, rewards = self.replay_memory.sample_batched(self.batch_size)
+        
+        # Reshape the states so they fit in the DQN model
+        states = states.view(self.batch_size, self.channels, self.grid_size, self.grid_size)
+        new_states = new_states.view(self.batch_size, self.channels, self.grid_size, self.grid_size)
+        
+        q_arrays = None
+        with torch.no_grad():
+            q_arrays = self.target_net(new_states)
+        q_max, _ = torch.max(q_arrays, axis=1, keepdim=True)
+        q_values = torch.multiply(q_max, self.gamma) + rewards.view(-1, 1)
+
+        self.optimizer.zero_grad()
+        self.policy_net.loss(states, actions, q_values).backward()
+
+        for param in self.policy_net.parameters():
+            param.grad.data.clamp(-1, 1)
+        self.optimizer.step()
+
+        self.steps += 1
+        if self.steps % self.T == 0:
+            self.__copy_q_to_target()
             
+        return observation, reward, terminated, truncated, info
+
+    def __memorize(self, state: torch.Tensor, new_state: torch.Tensor, action: int, reward: int): 
+        self.replay_memory.push(state, new_state, action, reward)
+
+    def __get_random_action(self):
+        return randint(0, 2)
+
+    def __get_action(self) -> int:
+        epsilon = self.epsilon_end \
+                    + (self.epsilon_start - self.epsilon_end) \
+                    * np.exp(-1 * self.steps / self.epsilon_decay)
+                    
+        if np.random.random() < epsilon:
+            return self.__get_random_action()
+        else:
+            action_q_vals = None
+            with torch.no_grad():
+                action_q_vals = self.policy_net(self.tensorized_obs.view(1, self.channels, self.grid_size, self.grid_size))
+            return torch.argmax(action_q_vals).item()
+
+    def __experience_replay(self, explore_only=False):
+        action = self.__get_action() if not explore_only else self.__get_random_action()
+
+        observation, reward, terminated, truncated, info = self._env_update(action)
+
+        new_observation = self.__tensorize_observation(observation)
+        self.__memorize(self.tensorized_obs, new_observation, action, reward)
+        self.tensorized_obs = new_observation
+        
+        return observation, reward, terminated, truncated, info
+
+    def __experience_initial(self):
+        for _ in range(20_000):
+            self.__experience_replay(explore_only=True)
+        self.tensorized_obs = self.__tensorize_observation(self.env.reset())
+
+    def __copy_q_to_target(self):
```

Comment:
as the q network is called policy network this should be called __copy_policy_to_target

Reply:
Another good spot - fixed!